File: C++/magic-squares-in-grid.cpp
// Time:  O(m * n)
// Space: O(1)

class Solution {
public:
    int numMagicSquaresInside(vector<vector<int>>& grid) {
        int result = 0;
        for (int i = 0; i + k - 1 < grid.size(); ++i) {
            for (int j = 0; j + k - 1 < grid[j].size(); ++j) {
                if (magic(grid, i, j)) {
                    ++result;
                }
            }
        }
        return result;
    }

private:
    static const int k = 3;

    bool magic(const vector<vector<int>>& grid, int r, int c) {
        const int expect = k * (k * k + 1) / 2; 
        unordered_set<int> nums;
        int min_num = numeric_limits<int>::max();
        int sum_diag = 0, sum_anti = 0;
        for (int i = 0; i < k; ++i) {
            sum_diag += grid[r + i][c + i];
            sum_anti += grid[r + i][c + k - 1 - i];
            int sum_r = 0, sum_c = 0;
            for (int j = 0; j < k; ++j) {
                min_num = min(min_num, grid[r + i][c + j]);
                nums.emplace(grid[r + i][c + j]);
                sum_r += grid[r + i][c + j];
                sum_c += grid[r + j][c + i];
            }
            if (!(sum_r == sum_c &&
                  sum_c == expect)) {
                return false;
            }
        }
        return sum_diag == sum_anti &&
               sum_anti == expect &&
               nums.size() == k * k &&
               min_num == 1;
    }
};

File: Python/magic-squares-in-grid.py
# Time:  O(m * n)
# Space: O(1)


class Solution(object):
    def numMagicSquaresInside(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        def magic(grid, r, c):
            expect = k * (k**2+1) // 2
            nums = set()
            min_num = float("inf")
            sum_diag, sum_anti = 0, 0
            for i in xrange(k):
                sum_diag += grid[r+i][c+i]
                sum_anti += grid[r+i][c+k-1-i]
                sum_r, sum_c = 0, 0
                for j in xrange(k):
                    min_num = min(min_num, grid[r+i][c+j])
                    nums.add(grid[r+i][c+j])
                    sum_r += grid[r+i][c+j]
                    sum_c += grid[r+j][c+i]
                if not (sum_r == sum_c == expect):
                    return False
            return sum_diag == sum_anti == expect and \
                len(nums) == k**2 and \
                min_num == 1

        k = 3
        result = 0
        for r in xrange(len(grid)-k+1):
            for c in xrange(len(grid[r])-k+1):
                if magic(grid, r, c):
                    result += 1
        return result

